File: C++/virtual-world-simulator/src/Entity/Plants/PineHogweed.cpp
#include "PineHogweed.h"

PineHogweed::PineHogweed(World* world, Point position) :
    Plant{ AssetManager::getAssetManager()->getAsset("pineHogweed"), world, position, 'P', 10, SpreadChance::LOW }
{

}

PineHogweed::PineHogweed(World* world, Point position, int lifeSpan, int strength) :
    Plant{ AssetManager::getAssetManager()->getAsset("pineHogweed"), world, position, 'P', strength, lifeSpan, SpreadChance::LOW, }
{
}

void PineHogweed::update(std::fstream& logFile)
{
    Entity*** map = world->getMap();
    int currentX = this->position.x;
    int currentY = this->position.y;

    for (int i = currentX - 1; i <= currentX + 1; i++)
    {
        if (i < 0 || i > world->getHeight() - 1)
            continue;

        for (int j = currentY - 1; j <= currentY + 1; j++)
        {
            if (j < 0 || j > world->getWidth() - 1)
                continue;

            if (map[i][j] != nullptr && map[i][j]->getPriority() != 0)
                map[i][j]->kill();
        }
    }
}

bool PineHogweed::collision(std::fstream& logFile, Entity& entity)
{
	logFile << this->position << " " << entity.getSymbol() << " was killed when eating " << this->symbol << '\n';
	entity.kill();
	this->kill();
	return true;
}

File: Python/virtual-world-simulator/src/Entity/Plants/PineHogweed.py
import pygame
from src.Entity.Plants.Plant import *


class PineHogweed(Plant):

    def __init__(self, world, position):
        temp = pygame.image.load('assets/pineHogweed.png')
        scaled = pygame.transform.scale(temp, (world.scale, world.scale))
        super().__init__(scaled, world, position, 'P', 10, SpreadChance.LOW)

    def update(self):
        currentX = self._position[0]
        currentY = self._position[1]

        for i in range(currentX - 1, currentX + 2):
            if i < 0 or i > self._world.height - 1:
                continue

            for j in range(currentY - 1, currentY + 2):
                if j < 0 or j > self._world.width - 1:
                    continue

                element = self._world.getMapElement(i, j)
                if element is not None and element.priority != 0:
                    element.kill()

        spreadPosition = self.spread()
        if spreadPosition is not None:
            element = PineHogweed(self._world, spreadPosition)
            self._world.setMapElement(spreadPosition[0], spreadPosition[1], element)
        super().update()

    def collision(self, entity):
        entity.kill()
        self.kill()
        return True